drgn_tools/meminfo.py

```diff
@@ -0,0 +1,755 @@
+from typing import List
+
+from drgn import Object
+from drgn import Program
+from drgn.helpers.linux import list_for_each_entry
+from drgn.helpers.linux.percpu import percpu_counter_sum
+
+from drgn_tools.corelens import CorelensModule
+from drgn_tools.mm import totalram_pages
+from drgn_tools.util import enum_value_get
+from drgn_tools.util import get_uts
+from drgn_tools.util import has_member
+
+
+__all__ = ("show_all_meminfo", "get_all_meminfo")
+
+
+# Machine info
+_IS_X86_64 = False
+_IS_AARCH_64 = False
+_UEK_VERSION = 5
+
+# Basic memory subsystem parameters are in the following:
+# The number of NUMA
+_NODES_SHIFT = 10
+_MAX_NUM_NODES = 1 << _NODES_SHIFT
+_MAX_NUM_ZONES = 0
+
+# Page settings.
+_PAGE_SIZE = 0
+_PAGE_SHIFT = 0
+_PMD_SHIFT = 0
+_HPAGE_SHIFT = 0
+
+_HPAGE_PMD_SHIFT = 0
+_HPAGE_PMD_ORDER = 0
+_HPAGE_PMD_NR = 0
+
+# vmalloc memory range.
+_VMALLOC_START = 0
+_VMALLOC_END = 0
+
+# Using five bits for the type means that the maximum number of
+# swapcache pages is 27 bits * on 32-bit-pgoff_t architectures.
+_MAX_SWAPFILES_SHIFT = 5
+# NUMA node memory migration support
+_SWP_DEVICE_NUM = 2
+# NUMA node memory migration support
+_SWP_MIGRATION_NUM = 2
+# Handling of hardware poisoned pages with memory corruption.
+_SWP_HWPOISON_NUM = 1
+
+# |MAX_SWAPFILES| defines the maximum number of swaptypes.
+_MAX_SWAPFILES = (
+    (1 << _MAX_SWAPFILES_SHIFT)
+    - _SWP_DEVICE_NUM
+    - _SWP_MIGRATION_NUM
+    - _SWP_HWPOISON_NUM
+)
+
+# The index for unknown stats ID for node stats and zone stats.
+UNKNOWN_GLOBAL_STAT_ID = -1
+
+
+def print_val_kb(text: str, num: int) -> None:
+    """
+    Produce the formatted output that matches the output in /proc/meminfo.
+
+    param text: The output item's name, e.g, "MemTotal"
+    param num: The output item's value
+    """
+    if num < 0:
+        return
+    text += ":"
+    num_kb = num << (_PAGE_SHIFT - 10)
+    print("%s%s kB" % (text.ljust(16, " "), str(num_kb).rjust(8, " ")))
+
+
+def parse_machine_info(prog: Program) -> None:
+    """
+    Parse the system's information to determine architecture-specific parameters.
+    This function must be called before parsing vmcore's memory statistics.
+    """
+    global _IS_X86_64, _IS_AARCH_64, _UEK_VERSION
+    global _MAX_NUM_NODES, _MAX_NUM_ZONES
+    global _PAGE_SIZE, _PAGE_SHIFT, _PMD_SHIFT, _HPAGE_SHIFT
+    global _HPAGE_PMD_SHIFT, _HPAGE_PMD_ORDER, _HPAGE_PMD_NR
+    global _VMALLOC_END, _VMALLOC_START, _VMALLOC_SPACE
+
+    uts = get_uts(prog)
+    # Set UEK version
+    if "4.14.35" in uts["release"]:
+        _UEK_VERSION = 5
+    elif "5.4.17" in uts["release"]:
+        _UEK_VERSION = 6
+    else:
+        _UEK_VERSION = 7
+
+    # Determine page size
+    _PAGE_SIZE = prog.constant("PAGE_SIZE").value_()
+    _PAGE_SHIFT = prog.constant("PAGE_SHIFT").value_()
+
+    # Set arch
+    if "aarch64" in uts["machine"]:
+        _IS_AARCH_64 = True
+
+        # include/linux/numa.h
+        _NODES_SHIFT = 2
```

Comment:
So, this is an excellent example of a constant that's really difficult to get. It's defined as a `#define`, and it comes from the kernel config. I'm actually working on a way to make it easy and reliable to access the kernel configuration from Drgn, but this will take a while, so we can't hope for that to work.

But we can still get this without hard-coding it. Notice that below the definition of `NODES_SHIFT`, we get `#define MAX_NUMNODES (1 << NODES_SHIFT)`. And `MAX_NUMNODES` is used in the definition of lots of data structures! I identified one (`preferred_node_policy`) which is defined in `mm/mempolicy.c` as:

```
static struct mempolicy preferred_node_policy[MAX_NUMNODES];
```

This is a pretty unique name and it seems to be present from UEK4 to mainline, so long as `CONFIG_NUMA` is enabled. We can use this to get the desired value:

```python
def nodes_shift(prog: Program) -> int:
    return int(math.log2(len(prog["preferred_node_policy"])))
```

Of course you'd want to include a comment describing why this is done, otherwise it will be incomprehensible to readers.

Reply:
Thanks for the hint. I've made changes to these constant variables.